programs/ziti-edge-tunnel/windows-scripts.c

```diff
@@ -0,0 +1,355 @@
+    FILE *fp;
+    char path[BUFFER_SIZE];
+
+    fp = popen(program, "r");
+    if (fp == NULL) {
+        ZITI_LOG(ERROR,"Failed to run command %s", strerror(errno));
+        return NULL;
+    }
+    char* result = calloc(MAXBUFFERLEN, sizeof(char));
+
+    while (fgets(path, sizeof(path), fp) != NULL) {
+        strcat(result, path);
+    }
+
+    pclose(fp);
+
+    return result;
+}
+
+void chunked_add_nrpt_rules(uv_loop_t *ziti_loop, LIST_HEAD(hostnames_list, hostname_s) *hostnames, char* tun_ip) {
+    char* script = calloc(MAX_POWERSHELL_SCRIPT_LEN, sizeof(char));
+    size_t buf_len = sprintf(script, "$Namespaces = @(");
+    size_t copied = buf_len;
+    int domains_size = 0;
+
+    while(!LIST_EMPTY(hostnames)) {
+        struct hostname_s *hostname = LIST_FIRST(hostnames);
+        buf_len = sprintf(script + copied, namespace_template, "\n", hostname->hostname);
+        copied += buf_len;
+        domains_size++;
+        LIST_REMOVE(hostname, _next);
+        free(hostname->hostname);
+        free(hostname);
+    }
+    buf_len = sprintf(script + copied, "%s\n\n", ")");
+    copied += buf_len;
+
+    buf_len = sprintf(script + copied, "ForEach ($Namespace in $Namespaces) {\n");
+    copied += buf_len;
+    buf_len = sprintf(script + copied, "$ns=$Namespace['n']\n");
+    copied += buf_len;
+    buf_len = sprintf(script + copied, "$Rule = @{Namespace=${ns}; NameServers=@('%s'); Comment='Added by %s'; DisplayName='%s:'+${ns}; }\n", tun_ip, exe_name, exe_name);
+    copied += buf_len;
+    buf_len = sprintf(script + copied, "Add-DnsClientNrptRule @Rule\n");
+    copied += buf_len;
+    buf_len = sprintf(script + copied, "}\n");
+    copied += buf_len;
+
+    ZITI_LOG(TRACE, "Adding NRPT script containing %d domains. total script size: %d", domains_size, copied);
+
+    char cmd[MAX_POWERSHELL_COMMAND_LEN];
+    snprintf(cmd, sizeof(cmd),"powershell -Command \"%s\"", script);
+
+    ZITI_LOG(INFO, "Executing ADD NRPT script :");
+    ZITI_LOG(INFO, "%s", cmd);
+    char* args[] = {"powershell", "-Command", script, NULL};
+    bool result = exec_process(ziti_loop, args[0], args);
+    if (!result) {
+        ZITI_LOG(WARN, "ADD NRPT script: %d(err=%d)", result, GetLastError());
+    }
+    free(script);
+}
+
+void add_nrpt_rules_script(uv_loop_t *nrpt_loop, struct add_service_nrpt_req *add_svc_req_data) {
+    model_map *hostnames = add_svc_req_data->hostnames;
+    char* dns_ip = add_svc_req_data->dns_ip;
+    if (hostnames == NULL || model_map_size(hostnames) == 0) {
+        ZITI_LOG(DEBUG, "No domains specified to add_nrpt_rules, exiting early");
+        return;
+    }
+    int namespace_template_padding = strlen(namespace_template);
+    LIST_HEAD(hostnames_list, hostname_s) host_names_list = LIST_HEAD_INITIALIZER(host_names_list);
+    int current_size = 0;
+    int rule_size = 0;
+    model_map_iter it = model_map_iterator(hostnames);
+    while(it != NULL) {
+        char* hostname = model_map_it_key(it);
```

Comment:
this should be `const char *` to match the return type. declaring the pointer with const prevents you from inadvertently changing the value.